### PR TITLE
allow dismissing return value of no-return pure functions

### PR DESCRIPTION
**Pull request by azjezz**
refs https://github.com/azjezz/psl/pull/163

no-return function either:
1. exit
2. throw exceptions
3. run forever

in all of these cases, it's okay not to use the return value.

### CODE REVIEW

**Comment by weirdan** [comment]
And you want to keep such functions pure to be able to call them in pure contexts, right?

**Comment by azjezz** [comment]
exactly.

**Comment by azjezz** [comment]
example

- `no-return`/`pure` : [invariant_violations](https://github.com/azjezz/psl/blob/a23f14859efebad0226b6c7cf4edb826e0ffa192/src/Psl/invariant_violation.php#L16)
- `pure`: [invariant](https://github.com/azjezz/psl/blob/a23f14859efebad0226b6c7cf4edb826e0ffa192/src/Psl/invariant.php#L21)
- `pure`: [log](https://github.com/azjezz/psl/blob/a23f14859efebad0226b6c7cf4edb826e0ffa192/src/Psl/Math/log.php#L18)

**Comment by azjezz** [comment]
currently to avoid this i suppressed all `UnusedFunctionCall` ( https://github.com/azjezz/psl/blob/1.6.x/psalm.xml#L65 ), but i don't really want to do that

**Comment by azjezz** [comment]
also relevant: https://github.com/azjezz/psl/pull/163#discussion_r599985666

**Comment by azjezz** [comment]
build failure is unrelated.

**Comment by muglug** [comment]
I think this is exactly the right approach, thanks!

While traditionally throwing an exception might be considered a side-effect, I think Psalm's notion of purity is safe enough

File: src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallAnalyzer.php [hunk 1]
```diff
@@ -1003,7 +1003,20 @@ private static function checkFunctionCallPurity(
...
+                 * it's okay to dismiss it's return value.
+                 */
+                if (
+                    !$context->inside_assignment &&
+                    !$context->inside_call &&
+                    !$context->inside_use &&
+                    !(
+                        $function_call_info->function_storage &&
+                        $function_call_info->function_storage->return_type &&
+                        (array_values($function_call_info->function_storage->return_type->getAtomicTypes())[0] ?? null) instanceof Type\Atomic\TNever
```
Comment (muglug) [inline src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallAnalyzer.php:1017]: I replaced this with `$function_call_info->function_storage->return_type->isNever()`

Comment (azjezz) [reply src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallAnalyzer.php:1017]: i didn't notice the method existed 😅

**Pull request merged**